### PR TITLE
Unified configuration

### PR DESCRIPTION
**Pull request by beam2d**
This PR implements a unified way of managing the global/thread-local configuration of Chainer. As a first step, it implements the core part of configuration (GlobalConfig and LocalConfig), and moves some easy configurations (debug mode, type check, and backprop mode) to this configuration tool. The main features are:

- We can easily add a new entry just by setting an attribute to `chainer.global_config`.
- `with chainer.using_config('XXX', yyy):` lets us change the thread-local configuration temporarily.
- We can diagnose the current configuration with `chainer.config.show()`.

I did not finish #2048 by this PR; train/test mode and use_cudnn flag are not moved to this configuration yet, because both affect a wide range of codes in Chainer and its examples, and I want to separate PRs for them. I will work on them after this PR being merged.

### CODE REVIEW

File: chainer/__init__.py [hunk 1]
```diff
@@ -101,8 +108,7 @@ def set_debug(debug):
     Args:
         debug (bool): New debug mode.
     """
-    global _debug
-    _debug = debug
+    config.debug = debug
 
 
 class DebugMode(object):
```
Comment (unnonouno) [inline chainer/__init__.py]: DebugMode is not required. How about making it deprecated?

File: chainer/configuration.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    def __getattr__(self, name):
+        if hasattr(self._local, name):
+            return getattr(self._local, name)
+        return getattr(self._global, name)
+
+    def __setattr__(self, name, value):
+        setattr(self._local, name, value)
+
+    def show(self, file=sys.stdout):
```
Comment (unnonouno) [inline chainer/configuration.py]: Do we need to make the same method in GlobalConfig?

File: chainer/__init__.py (hunk 1 above)
Comment (beam2d) [inline chainer/__init__.py]: OK, I will.

File: chainer/configuration.py (hunk 2 above)
Comment (beam2d) [inline chainer/configuration.py]: Ah, yes. I will.

**Comment by beam2d** [comment]
@unnonouno Resolved the comments.

**Comment by unnonouno** [comment]
LGTM

**Comment by unnonouno** [comment]
Check flake8 result.

**Comment by beam2d** [comment]
@unnonouno fixed relevant codes for flake8.

**Comment by unnonouno** [comment]
Please check travis result.

**Comment by beam2d** [comment]
Sorry, I re-fixed it. I think the code format issue is resolved also for Travis.

**Comment by unnonouno** [comment]
LGTM!

**Pull request merged**